### PR TITLE
Add outdated version warning to docs

### LINKED ISSUE #1599: Show warning banner when viewing documentation for old versions
### Is your feature request related to a problem? Please describe.

It has happened multiple times that I've landed at the official K0S documentation via Google etc., and ended up at the correct docs page, but at an older version. 

### Describe the solution you would like

Since the docs are using the [MkDocs-Material](https://github.com/squidfunk/mkdocs-material) theme, they seem to already provide a solution for this problem:

https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning/#version-warning

image

### Describe alternatives you've considered

_No response_

### Additional context

One example of the first Google result pointing to an outdated version:

image

### PR DESCRIPTION
**Pull request by ruhrscholz**
## Description

Fixes #1599. This is an exact copy of the example code provided in https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning/#version-warning, although I could not get the banner to actually show in my local builds. The provided build scripts are not set up for Kine use.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation update

## How Has This Been Tested?

- [X] Manual test
- [ ] Auto test added

## Checklist:

- [X] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [X] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by twz123** [comment]
Cool, thanks! Could you also set the default version to stable, as described in the link you posted? What happens if you browse to some pre-release or the head docs? Will the warning be displayed there, as well?

**Comment by ruhrscholz** [comment]
> Cool, thanks! Could you also set the default version to stable, as described in the link you posted? What happens if you browse to some pre-release or the head docs? Will the warning be displayed there, as well?

I set the default version to stable and updated the PR. My problem is that I currently cannot verify correct behavior, as the GitHub workflow to publish the docs involves some custom steps that I could not reproduce yet.

**Comment by twz123** [comment]
I usually test this with some slight modifications to the workflow. Something along those lines: d6ecdd67283d8b7c7c1ca248f97bdd2c7677ea59

No worries if you don't get it to work, I'll have a look then.

**Comment by ruhrscholz** [comment]
The good news: I could get it to work. You can view it for some versions here: https://ruhrscholz.github.io/k0s/. Please not that I did not redeploy the docs for each and every version as the mkdocs version and the banner did not change.

The bad news: It would involve quite a bit of work to the current repo. The problem is that this "outdated" snippet was designed to be included since version v0.0.1. There seem to be multiple issues:

1. The older versions need to be rebuild, either manually or automatically
2. The older versions don't contain the necessary `outdated` snipped: This should be easily solvable since one can just overwrite the old `docs/custom_theme/main.html` with the new one from the PR. There don't seem to have been any changes to it since that file was introduced (around v0.12 IIRC; example from my test Workflow: https://github.com/ruhrscholz/k0s/commit/d0885ac34279bc6c64e08288a4652c24bb86702b?diff=split)
3. The current (manual) build process is broken. Especially the Makefile contains an incorrect link for the GitHub API: https://github.com/ruhrscholz/k0s/commit/c20e6d625faee5b3a774447c6b8d1295781aef64?diff=split

All in all, there were quite a few things I had to change in the Manual Docs Deploy Workflow to even get it to build: https://github.com/ruhrscholz/k0s/commit/2d83d781083cecac3d639ff7b8310f739c0c8f08 (the `ruhrscholz-patch-1` branch is just to test the custom deploy script). I also had to specify the `stable` version in the manual deployment script, this seems to be automated in the normal script.

What has to be done:
- Fix the bug in the Makefile
- Create a custom deploy workflow that overrides the `main.html` with the modified one. Since this is a one-off action this could be done locally instead of in GitHub Actions.
- It may be possible that `publish-docs.yml` has to be fixed afterwards, although in theory it should work fine.

> What happens if you browse to some pre-release or the head docs? Will the warning be displayed there, as well?

No, pre-release versions do not contain the Banner: https://ruhrscholz.github.io/k0s/v1.23.4-rc.1+k0s.0/

I will try to write a quick bash script to redeploy every docs version in the next days if the maintainers want to continue in adding this banner at all.

**Comment by ruhrscholz** [comment]
A quick status update: I wrote a Python script to regenerate all docs. This took some trial and error since there are some edge-cases in the Makefile and the GitHub workflows. You can find the current version here: https://gist.github.com/ruhrscholz/2615c7561b4b55f14925b16d6f532c0f

Currently I'm waiting for it to finish building, this takes some time since it has to re-download all old K0S versions to generate the CLI documentation.

Another caveat was parsing all old `mkdocs.yml` files correctly to append the configuration for the outdated block. I could not use the current `mkdocs.yml` for that since it contains references to the navigation, which changed between versions.

After the script has run through I'll compare the output with the HEAD of the `gh-pages` branch to check for any more edge cases. If it looks fine I'll probably create another PR to merge the new `gh-pages` into the current one.

**Comment by twz123** [comment]
> The good news: I could get it to work. You can view it for some versions here: https://ruhrscholz.github.io/k0s/. Please not that I did not redeploy the docs for each and every version as the mkdocs version and the banner did not change.

Very cool!
 
> The bad news: It would involve quite a bit of work to the current repo. The problem is that this "outdated" snippet was designed to be included since version v0.0.1. There seem to be multiple issues:
> 
>     1. The older versions need to be rebuild, either manually or automatically
> 
>     2. The older versions don't contain the necessary `outdated` snipped: This should be easily solvable since one can just overwrite the old `docs/custom_theme/main.html` with the new one from the PR. There don't seem to have been any changes to it since that file was introduced (around v0.12 IIRC; example from my test Workflow: [ruhrscholz@d0885ac?diff=split](https://github.com/ruhrscholz/k0s/commit/d0885ac34279bc6c64e08288a4652c24bb86702b?diff=split))

As a start, let's just focus on the current docs versions. Let's land the outdated notice there, and then let's see if it's feasible to rebuild the older docs with that notice as well. We also had issues with the version picker (#1578). So regenerating the older docs could cover that as well.

>     3. The current (manual) build process is broken. Especially the Makefile contains an incorrect link for the GitHub API: [ruhrscholz@c20e6d6?diff=split](https://github.com/ruhrscholz/k0s/commit/c20e6d625faee5b3a774447c6b8d1295781aef64?diff=split)

That should have been addressed in [#1564](https://github.com/k0sproject/k0s/pull/1564/files?diff=unified&w=0#diff-cdec797dddf8fc387304501d6bd4f318e1e9067c6fae012dfa8b69fd85fd3248R29-R54).

> No, pre-release versions do not contain the Banner: https://ruhrscholz.github.io/k0s/v1.23.4-rc.1+k0s.0/

That link doesn't resolve any more, but I remember that I saw the outdated banner there, last week. I also had a look at the material-docs theme, and If I understood it correctly, there's only a single version that can suppress the banner. That's why I'd propose a slightly different wording for it.

File: docs/custom_theme/main.html [hunk 1]
```diff
@@ -10,4 +10,11 @@
 
     gtag('config', 'G-JMV1409T2T');
   </script>
+{% endblock %}
+
+{% block outdated %}
+  You're not viewing the latest version.
```
Comment (twz123) [inline docs/custom_theme/main.html:16]:
```suggestion
  You are not currently viewing the documentation for the current stable release of k0s.
```

**Review by twz123 — CHANGES_REQUESTED** [review]

**Comment by ruhrscholz** [comment]
> That should have been addressed in [#1564](https://github.com/k0sproject/k0s/pull/1564/files?diff=unified&w=0#diff-cdec797dddf8fc387304501d6bd4f318e1e9067c6fae012dfa8b69fd85fd3248R29-R54).

You are correct, I just tested the build script from the PR.

> That link doesn't resolve any more,

Yes, I rebuilt the docs accessible from that link in PR #1602. This follows the at the time existing build script that skipped beta and RC builds.

> there's only a single version that can suppress the banner. That's why I'd propose a slightly different wording for it.

I completely agree, I've commited your changes.

Is there anything left to do for me?

**Comment by twz123** [comment]
Thanks for adding the sign-off.

**Review by twz123 — APPROVED** [review]

**Comment by twz123** [comment]
@mviitane @jnummelin Are you okay with the current wording of the banner?

**Comment by jnummelin** [comment]
Sounds good to me

**Comment by twz123** [comment]
Thanks @ruhrscholz.

**Comment by twz123** [comment]
Result for https://docs.k0sproject.io/head/

Screenshot 2022-03-30 at 14-48-12 Documentation

**Pull request merged**